915 implement person model

mononoken opened the pull request:
# 🔗 Issue
#915 

# ✍️ Description

Basically, this removes AdopterFosterAccount and wires up Person to be used in its place wherever necessary.

If you have any questions, please let me know, or we can discuss more in the next meeting.

I fixed some unrelated bugs and such and cleaned up some things while making this refactoring as well. I also fixed some behavior that wasn't working on main where I could. For example, adopters should be able to submit adoptions now (if they have a form submission, which they can't currently create themselves but we can make in console).

# 📷 Screenshots/Demos

None. Everything at the top level should mostly behave as it did before.

mononoken commented:
Not sure what is going on with the failing tests. I'll try to take a look later this week.

mononoken commented:
Getting some flakey test behavior on my machine which think is reflected on main too. Not sure if I will have time to investigate it until the weekend.

kasugaijin commented:
@mononoken I should get to take a look this weekend! Thanks for this PR.

mononoken commented:
> @mononoken I should get to take a look this weekend! Thanks for this PR.

Let me know if you have any questions or find any mistakes!

kasugaijin commented on app/controllers/organizations/adoptable_pets_controller.rb:
```diff
@@ -21,14 +22,14 @@ def show
-          pet_id: @pet.id,
-          adopter_foster_account_id: current_user.adopter_foster_account.id
+          pet: @pet,
+          form_submission: current_user.person.form_submission
```
I was thinking we would need to call `person.form_submission.last` because we would allow for multiple form submissions/sets of answers. But as I think of it, perhaps it makes sense to just keep one set (the latest set) of answers as the 'truth'. There's no point in keeping record of historical answers, right? We currently declare `has_one :form_submission` on `Person` so no need to change that in this case. Just raising here for a thought.

I was originally thinking we would track all form submissions i.e., keep the answers every time a form is submitted for a given person, and just just the last one for Matches. But, I cannot think of a compelling reason to have multiple at this stage. Can you?

kasugaijin commented on app/controllers/organizations/staff/matches_controller.rb:
```diff
@@ -32,7 +32,7 @@ def destroy
 
   def match_params
-    params.require(:match).permit(:pet_id, :adopter_foster_account_id, :match_type, :start_date, :end_date)
+    params.require(:match).permit(:pet_id, :match_type, :start_date, :end_date)
```
Does this need a `:person_id` ?

mononoken replied:
Added 85edebb507fd

mononoken replied:
This commit also fixes creating adoptions

kasugaijin commented on app/mailers/adoption_mailer.rb:
```diff
@@ -1,11 +1,8 @@
     @pet = match.pet
 
-    mail to: @user.email, subject: "#{@pet.name}'s Reminder"
+    mail to: person.email, subject: "#{@pet.name}'s Reminder"
```
We don't need the person instance variable in the view template do we?

kasugaijin commented on app/models/adopter_application.rb:
Do we want to add similar validations here to prevent a Person from applying on a pet if they have already done so?

kasugaijin replied:
Ah I see there is a check in the policy as well. Might still be best to keep a model level check just in case?

mononoken replied:
That sounds right but think it is not in the scope of this PR. Should be a fairly easy issue. Would just need to add a validation, and it would be good to create a unique index on the postgres server too.

kasugaijin commented:
@mononoken looking good. I have left a few questions and comments. 
One key thing we will need to do is create the Person's form submission when the account is created. You're working on this right, @MooseCowBear ?

MooseCowBear commented:
@kasugaijin @mononoken Taking care of this. As of now, I'm creating a form submission like: 

```
def create
    super do |resource|
      if resource.persisted?
        resource.add_role(:adopter, Current.organization)
        ::Organizations::FormSubmissionService.new(resource).create
      end
    end
  end
```
because the form submission belongs to person and person is being created in a callback for user. Not sure that's the best place for it. We could create it after person is created in the callback as well?

kasugaijin commented:
> @kasugaijin @mononoken Taking care of this. As of now, I'm creating a form submission like:
> 
> ```
> def create
>     super do |resource|
>       if resource.persisted?
>         resource.add_role(:adopter, Current.organization)
>         ::Organizations::FormSubmissionService.new(resource).create
>       end
>     end
>   end
> ```
> 
> because the form submission belongs to person and person is being created in a callback for user. Not sure that's the best place for it. We could create it after person is created in the callback as well?

Ah yes good call. I think we are at risk of an issue here if the Person does not get created before we try to create the FormSubmission. You are right. I think we should instead have an after_create callback on Person that will create their form submission.

mononoken commented:
@kasugaijin , again, thank you for reviewing this one! I think I addressed all your comments. I fixed the adoption creation from reviews as well as the sign up error too.

kasugaijin reviewed (approved):
LGTM!
